File: src/lib/mcp.ts
/**
 * MCP (Model Context Protocol) client helper.
 *
 * Mirrors the serialized shapes from `crates/ha-mcp/src/api.rs`
 * and `ha-config-schema/src/mcp.rs`. Every function here is a
 * thin wrapper over `getTransport().call(...)` so switching between
 * Tauri IPC and HTTP transport is invisible to callers.
 */

import { getTransport } from "@/lib/transport-provider";

// ── Enums & unions ───────────────────────────────────────────────

export type McpTransportKind = "stdio" | "streamableHttp" | "sse" | "websocket";

export type McpTransportSpec =
  | { kind: "stdio"; command: string; args?: string[]; cwd?: string | null }
  | { kind: "streamableHttp"; url: string }
  | { kind: "sse"; url: string }
  | { kind: "websocket"; url: string };

export type McpTrustLevel = "untrusted" | "trusted";

/** Matches `crates/ha-mcp/src/registry.rs::ServerState::label()`. */
export type McpServerState =
  | "disabled"
  | "idle"
  | "connecting"
  | "ready"
  | "needsAuth"
  | "failed";

// ── Persisted config shape ───────────────────────────────────────

export interface McpOAuthConfig {
  clientId?: string | null;
  clientSecret?: string | null;
  authorizationEndpoint?: string | null;
  tokenEndpoint?: string | null;
  scopes?: string[];
  extraParams?: Record<string, string>;
}

export interface McpServerConfig {
  id: string;
  name: string;
  enabled: boolean;
  transport: McpTransportSpec;
  env?: Record<string, string>;
  headers?: Record<string, string>;
  oauth?: McpOAuthConfig | null;
  allowedTools?: string[];
  deniedTools?: string[];
  connectTimeoutSecs: number;
  callTimeoutSecs: number;
  healthCheckIntervalSecs: number;
  maxConcurrentCalls: number;
  autoApprove: boolean;
  trustLevel: McpTrustLevel;
  eager: boolean;
  deferredTools?: boolean;
  projectPaths?: string[];
  description?: string | null;
  icon?: string | null;
  createdAt: number;
  updatedAt: number;
  trustAcknowledgedAt?: string | null;
}

/** Shape sent to `mcp_add_server` / `mcp_update_server` — id optional. */
export type McpServerDraft = Omit<
  McpServerConfig,
  | "id"
  | "createdAt"
  | "updatedAt"
  | "connectTimeoutSecs"
  | "callTimeoutSecs"
  | "healthCheckIntervalSecs"
  | "maxConcurrentCalls"
> & {
  id?: string;
  connectTimeoutSecs?: number | null;
  callTimeoutSecs?: number | null;
  healthCheckIntervalSecs?: number | null;
  maxConcurrentCalls?: number | null;
};

export interface McpGlobalSettings {
  enabled: boolean;
  maxConcurrentCalls: number;
  backoffInitialSecs: number;
  backoffMaxSecs: number;
  consecutiveFailureCircuitBreaker: number;
  autoReconnectAfterCircuitSecs: number;
  deniedServers?: string[];
}

// ── Runtime snapshots ────────────────────────────────────────────

export interface McpServerStatusSnapshot {
  id: string;
  name: string;
  enabled: boolean;
  transportKind: McpTransportKind;
  state: McpServerState;
  reason?: string | null;
  toolCount: number;
  resourceCount: number;
  promptCount: number;
  consecutiveFailures: number;
  lastHealthCheckTs: number;
}

/** `McpServerSummary` from api.rs — config + status merged. */
export type McpServerSummary = McpServerConfig & Omit<McpServerStatusSnapshot, "transportKind">;

export interface McpToolSummary {
  name: string;
  namespacedName: string;
  description?: string;
}

export interface McpLogLine {
  ts: number;
  level: string;
  source: string;
  message: string;
}

export interface McpImportSummary {
  imported: string[];
  skipped: { name: string; reason: string }[];
}

// ── Transport helpers ────────────────────────────────────────────

export function listServers(): Promise<McpServerSummary[]> {
  return getTransport().call("mcp_list_servers");
}

export function getServerStatus(id: string): Promise<McpServerStatusSnapshot> {
  return getTransport().call("mcp_get_server_status", { id });
}

export function addServer(draft: McpServerDraft): Promise<McpServerSummary> {
  return getTransport().call("mcp_add_server", { draft });
}

export function updateServer(
  id: string,
  draft: McpServerDraft,
): Promise<McpServerSummary> {
  return getTransport().call("mcp_update_server", { id, draft });
}

export function removeServer(id: string): Promise<void> {
  return getTransport().call("mcp_remove_server", { id });
}

export function reorderServers(order: string[]): Promise<void> {
  return getTransport().call("mcp_reorder_servers", { order });
}

export function testConnection(id: string): Promise<McpServerStatusSnapshot> {
  return getTransport().call("mcp_test_connection", { id });
}

export function reconnectServer(id: string): Promise<McpServerStatusSnapshot> {
  return getTransport().call("mcp_reconnect_server", { id });
}

export function startOauth(id: string): Promise<void> {
  return getTransport().call("mcp_start_oauth", { id });
}

export function signOut(id: string): Promise<void> {
  return getTransport().call("mcp_sign_out", { id });
}

export function listServerTools(id: string): Promise<McpToolSummary[]> {
  return getTransport().call("mcp_list_tools", { id });
}

export function getRecentLogs(
  id: string,
  limit: number = 200,
): Promise<McpLogLine[]> {
  return getTransport().call("mcp_get_recent_logs", { id, limit });
}

export function importClaudeDesktopConfig(
  json: string,
): Promise<McpImportSummary> {
  return getTransport().call("mcp_import_claude_desktop_config", { json });
}

export function getGlobalSettings(): Promise<McpGlobalSettings> {
  return getTransport().call("mcp_get_global_settings");
}

export function updateGlobalSettings(
  settings: McpGlobalSettings,
): Promise<void> {
  return getTransport().call("mcp_update_global_settings", { settings });
}

// ── EventBus subscriptions ──────────────────────────────────────

export const MCP_EVENTS = {
  SERVER_STATUS_CHANGED: "mcp:server_status_changed",
  CATALOG_REFRESHED: "mcp:catalog_refreshed",
  AUTH_REQUIRED: "mcp:auth_required",
  AUTH_COMPLETED: "mcp:auth_completed",
  SERVERS_CHANGED: "mcp:servers_changed",
  SERVER_LOG: "mcp:server_log",
} as const;

// ── UI helpers ──────────────────────────────────────────────────

/** Decompose a namespaced tool name into its constituents, or null when
 * the name isn't MCP-owned. */
export function parseMcpToolName(
  name: string,
): { serverName: string; tool: string } | null {
  if (!name.startsWith("mcp__")) return null;
  const rest = name.slice(5);
  const sep = rest.indexOf("__");
  if (sep <= 0) return null;
  return {
    serverName: rest.slice(0, sep),
    tool: rest.slice(sep + 2),
  };
}

File: src-tauri/src/commands/mcp.rs
//! Tauri IPC commands for the MCP subsystem.
//!
//! Thin shells over the business logic in
//! [`ha_mcp::api`]. Every command returns `Result<T, CmdError>` —
//! Tauri serializes `T` to JS via serde and surfaces the `Err` (rendered as
//! a string) to `invoke()` callers as a rejected promise.
//!
//! Both the Tauri invoke handler in `src-tauri/src/lib.rs` and the
//! matching axum routes in `crates/ha-server/src/routes/mcp.rs` dispatch
//! to the SAME `ha_mcp::api::*` functions — behavior parity is
//! enforced by the single source of truth, not by copy-paste.

use crate::commands::CmdError;
use ha_mcp::api::{
    self, ImportSummary, McpLogLine, McpServerDraft, McpServerSummary, McpToolSummary,
};
use ha_mcp::config::McpGlobalSettings;
use ha_mcp::registry::ServerStatusSnapshot;

use crate::AppState;
use tauri::State;

// ── CRUD ─────────────────────────────────────────────────────────

#[tauri::command]
pub async fn mcp_list_servers(
    _state: State<'_, AppState>,
) -> Result<Vec<McpServerSummary>, CmdError> {
    Ok(api::list_servers().await)
}

#[tauri::command]
pub async fn mcp_get_server_status(
    id: String,
    _state: State<'_, AppState>,
) -> Result<ServerStatusSnapshot, CmdError> {
    api::get_server_status(&id)
        .await
        .ok_or_else(|| CmdError::msg(format!("MCP server '{id}' not found")))
}

#[tauri::command]
pub async fn mcp_add_server(
    draft: McpServerDraft,
    _state: State<'_, AppState>,
) -> Result<McpServerSummary, CmdError> {
    api::add_server(draft).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_update_server(
    id: String,
    draft: McpServerDraft,
    _state: State<'_, AppState>,
) -> Result<McpServerSummary, CmdError> {
    api::update_server(&id, draft).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_remove_server(id: String, _state: State<'_, AppState>) -> Result<(), CmdError> {
    api::remove_server(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_reorder_servers(
    order: Vec<String>,
    _state: State<'_, AppState>,
) -> Result<(), CmdError> {
    api::reorder_servers(order).await.map_err(Into::into)
}

// ── Connection + diagnostics ─────────────────────────────────────

#[tauri::command]
pub async fn mcp_test_connection(
    id: String,
    _state: State<'_, AppState>,
) -> Result<ServerStatusSnapshot, CmdError> {
    api::test_connection(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_reconnect_server(
    id: String,
    _state: State<'_, AppState>,
) -> Result<ServerStatusSnapshot, CmdError> {
    api::reconnect_server(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_start_oauth(id: String, _state: State<'_, AppState>) -> Result<(), CmdError> {
    api::start_oauth(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_sign_out(id: String, _state: State<'_, AppState>) -> Result<(), CmdError> {
    api::sign_out(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_list_tools(
    id: String,
    _state: State<'_, AppState>,
) -> Result<Vec<McpToolSummary>, CmdError> {
    api::list_server_tools(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_get_recent_logs(
    id: String,
    limit: Option<usize>,
    _state: State<'_, AppState>,
) -> Result<Vec<McpLogLine>, CmdError> {
    api::get_recent_logs(&id, limit.unwrap_or(200))
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub async fn mcp_import_claude_desktop_config(
    json: String,
    _state: State<'_, AppState>,
) -> Result<ImportSummary, CmdError> {
    api::import_claude_desktop_config(&json)
        .await
        .map_err(Into::into)
}

// ── Global settings ──────────────────────────────────────────────

#[tauri::command]
pub async fn mcp_get_global_settings(
    _state: State<'_, AppState>,
) -> Result<McpGlobalSettings, CmdError> {
    Ok(api::get_global_settings())
}

#[tauri::command]
pub async fn mcp_update_global_settings(
    settings: McpGlobalSettings,
    _state: State<'_, AppState>,
) -> Result<(), CmdError> {
    api::update_global_settings(settings)
        .await
        .map_err(Into::into)
}
